### ISSUE TITLE
Can not handle uppercase keys, such as alt+shift+h

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#3382]
From @aeiouaoeiuv in https://github.com/microsoft/vscode/issues/126442

Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.57.0
- OS Version: Ubuntu 20.04 (5.4.0-42-generic)
- Bash Version: 5.0.17(1)-release
- Zsh Version: 5.8

Steps to Reproduce:

1. Open ubuntu terminal and start vscode: `code --disable-extensions`.
2. Open vscode terminal with `bash` or `zsh`.
3. Input `cat` command and enter.
4. Press `alt+shift+h` keys, and it shows up `^[h`.

What I expected to see is `^[H` with `alt+shift+h` pressed.
And ubuntu terminal shows the correct keys codes.

This is `ubuntu terminal` reproduce picture:
![ubuntu_terminal](https://user-images.githubusercontent.com/18081173/122165312-9d139300-ceaa-11eb-8b39-86483ff05a41.png)

This is `vscode terminal` reproduce picture:
![vscode_terminal](https://user-images.githubusercontent.com/18081173/122165318-a270dd80-ceaa-11eb-867d-6613cfe8fd61.png)

**Comment by Tyriar** [comment]
@jerch seems like a valid bug, right?

**Comment by jerch** [comment]
@Tyriar Yes thats off, yet another bug in key handling. We really should get down to reworking `Keyboard.ts`.

As you might remember I started a PR several months back, but abandoned it due to the route it took - basically we cannot expose plain key events anymore, but would have to synthesize our keystroke events from several low level ones. As this was not really testable other than by manually pressing keys on the keyboard, I got scared away. For a good refactoring we'd need a proper test solution first, no clue how we get that for all platforms **x** browsers **x** keyboard layouts **x** IME helper and such. Really a nightmare to touch.

**Comment by Tyriar** [comment]
So for this specific one is it not as easy as just checking `shiftKey` on the event?

**Comment by jerch** [comment]
> So for this specific one is it not as easy as just checking `shiftKey` on the event?

Yes, somewhere down the altKey branches we dont respect the shiftKey state. Did only a quick peek, looks like the else path in `keyCode` here: https://github.com/xtermjs/xterm.js/blob/094bcbd81d1f080d204331088aa99f4d7173d56c/src/common/input/Keyboard.ts#L355-L358

**Comment by aeiouaoeiuv** [comment]
```
Version: 1.100.2
Commit: 848b80aeb52026648a8ff9f7c45a9b0a80641e2e
Date: 2025-05-14T21:47:40.416Z
Electron: 34.5.1
ElectronBuildId: 11369351
Chromium: 132.0.6834.210
Node.js: 20.19.0
V8: 13.2.152.41-electron.0
OS: Linux x64 6.11.0-19-generic
```
This vscode version is already worked for alt+shift+h command

**Comment by Tyriar** [comment]
Apparently this works now: https://github.com/xtermjs/xterm.js/pull/5349

**Comment by jerch** [comment]
@Tyriar I love bugs, that just disappear. Weak determinism for P.. oh wait  😮
